devices/uart-nrf9160/uart.c

```diff
@@ -0,0 +1,323 @@
+	{ (void *)UART3_BASE, UART3_IRQ, UART_BAUDRATE, (volatile char *)UART3_TX_DMA, (volatile char *)UART3_RX_DMA,
+		UART3_TX, UART3_RX, UART3_RTS, UART3_CTS }
+};
+
+
+static uart_t *uart_getInstance(unsigned int minor)
+{
+	if (minor >= UART_MAX_CNT)
+		return NULL;
+
+	if (uartLut[minor] == 0)
+		return NULL;
+
+	return &uart_common.uarts[minor];
+}
+
+
+static int uart_handleIntr(unsigned int irq, void *buff)
+{
+	uart_t *uart = (uart_t *)buff;
+
+	if (uart == NULL)
+		return -EINVAL;
+
+	if (uart->base + uarte_events_rxdrdy) {
+		/* clear rxdrdy event flag */
+		*(uart->base + uarte_events_rxdrdy) = 0u;
+		/* we discard volatile type intentionally */
+		lib_cbufWrite(&uart->cbuffRx, &((char *)uart->rx_dma)[uart->cnt], 1);
+		uart->cnt++;
+	}
+	if (uart->base + uarte_events_endrx) {
+		/* clear endrx event flag */
+		*(uart->base + uarte_events_endrx) = 0u;
+		uart->cnt = 0;
+		*(uart->base + uarte_startrx) = 1u;
+		hal_cpuDataMemoryBarrier();
+	}
+
+	return 0;
+}
+
+/* Device interface */
+
+static ssize_t uart_read(unsigned int minor, addr_t offs, void *buff, size_t len, time_t timeout)
+{
+	size_t res;
+	uart_t *uart;
+	time_t start;
+
+	if ((uart = uart_getInstance(minor)) == NULL)
+		return -EINVAL;
+
+	start = hal_timerGet();
+	while (lib_cbufEmpty(&uart->cbuffRx)) {
+		if ((hal_timerGet() - start) >= timeout)
+			return -ETIME;
+	}
+
+	hal_interruptsDisable();
+	res = lib_cbufRead(&uart->cbuffRx, buff, len);
+	hal_interruptsEnable();
+
+	return res;
+}
+
+
+static size_t uart_send(uart_t *uart, size_t len)
+{
+	*(uart->base + uarte_txd_maxcnt) = len;
+	*(uart->base + uarte_starttx) = 1u;
+	while (*(uart->base + uarte_events_txstarted) != 1u)
+		;
+	*(uart->base + uarte_events_txstarted) = 0u;
+
+	while (*(uart->base + uarte_events_endtx) != 1u)
+		;
+	*(uart->base + uarte_events_endtx) = 0u;
+
+	return (size_t) * (uart->base + uarte_txd_amount);
+}
+
+
+static ssize_t uart_write(unsigned int minor, const void *buff, size_t len)
+{
+	uart_t *uart;
+	size_t cnt;
+
+	if ((uart = uart_getInstance(minor)) == NULL)
+		return -EINVAL;
+
+	hal_memcpy((void *)uart->tx_dma, buff, len);
+
+	cnt = uart_send(uart, len);
+
+	return (ssize_t)cnt;
+}
+
+
+static ssize_t uart_safeWrite(unsigned int minor, addr_t offs, const void *buff, size_t len)
+{
+	return uart_write(minor, buff, len);
+}
+
+
+static int uart_sync(unsigned int minor)
+{
+	uart_t *uart;
+
+	if ((uart = uart_getInstance(minor)) == NULL)
+		return -EINVAL;
+
+	while (!lib_cbufEmpty(&uart->cbuffTx))
+		;
+
+	/* endtx flag is asserted in uart_send so mustn't be checked here */
+
+	return EOK;
+}
+
+
+static int uart_done(unsigned int minor)
+{
+	uart_t *uart;
+
+	if ((uart = uart_getInstance(minor)) == NULL)
+		return -EINVAL;
+
+	/* Wait for transmission activity complete */
+	(void)uart_sync(minor);
+
+	*(uart->base + uarte_stoptx) = 1u;
+	*(uart->base + uarte_stoprx) = 1u;
+	/* disable all uart interrupts */
+	*(uart->base + uarte_intenclr) = 0xFFFFFFFF;
```

Comment:
This is one place you really want `u` suffix :D
```suggestion
	*(uart->base + uarte_intenclr) = 0xffffffffu;
```